node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,207 @@
+//! BLS signature scheme for the BN254 curve.
+
+use crate::ByteFmt;
+use anyhow::anyhow;
+use ark_bn254::{Bn254, Fr, G1Projective as G1, G2Projective as G2};
+use ark_ec::{
+    pairing::{Pairing as _, PairingOutput},
+    Group as _,
+};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+pub use error::Error;
+use num_traits::Zero as _;
+use std::collections::HashMap;
+
+#[doc(hidden)]
+pub mod error;
+
+#[cfg(test)]
+mod tests;
+
+pub mod hash;
+mod testonly;
+
+/// Type safety wrapper around a scalar value.
+pub struct SecretKey(Fr);
+
+impl SecretKey {
+    /// Gets the corresponding [`PublicKey`] for this [`SecretKey`]
+    pub fn public(&self) -> PublicKey {
+        let p = G2::generator() * self.0;
+        PublicKey(p)
+    }
+
+    /// Produces a signature using this [`SecretKey`]
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        let hash_point = hash::hash_to_g1(msg);
+        let sig = hash_point * self.0;
+        Signature(sig)
+    }
+}
+
+impl ByteFmt for SecretKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        Fr::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode secret key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+/// Type safety wrapper around G2.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct PublicKey(G2);
+
+impl ByteFmt for PublicKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G2::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode public key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+/// Type safety wrapper around a G1 value.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Signature(G1);
+
+impl Signature {
+    /// Verifies a signature against the provided public key
+    #[inline(never)]
```

Comment:
add a comment explaining why we have this nin-generic function and why inline never

Reply:
Added.